### PR TITLE
Client tool filtering

### PR DESCRIPTION
**Pull request by Unraveler**
Available client tools can now be controlled through client config.

Please review @terrestris/devs

### CODE REVIEW

**Review by dnlkoch — CHANGES_REQUESTED** [review]
In my understanding of the client/SHOGun, all application specific configurations, e.g. for the view, the layers or even the visible tools, should be customizable for each application individually rather than global.

File: src/index.tsx [hunk 1]
```diff
@@ -173,6 +177,19 @@ const setApplicationToStore = async (application?: Application) => {
...
+  if (application.toolConfig) {
+    const availableTools: string[] = [];
+    application.toolConfig
+      .map((tool: DefaultApplicationToolConfig) => {
+        if (tool.config.visible) {
+          availableTools.push(tool.name);
+        };
+      });
+    store.dispatch(setSelectedKey(availableTools)
+    );
```
Comment (dnlkoch) [inline src/index.tsx:190]:
```suggestion
    store.dispatch(setSelectedKey(availableTools));
```

File: src/global.d.ts [hunk 2]
```diff
@@ -14,6 +14,28 @@ declare module 'clientConfig' {
...
+        modify?: boolean;
+        download?: boolean;
+        upload?: boolean;
+        delete?: boolean;
+      };
+      feature_info?: boolean;
+      print?: boolean;
+      tree?: boolean;
+      permalink?: boolean;
+    };
```
Comment (dnlkoch) [inline src/global.d.ts:38]: Is this still needed?

File: src/index.tsx [hunk 3]
```diff
@@ -173,6 +177,19 @@ const setApplicationToStore = async (application?: Application) => {
...
+
+  if (application.toolConfig) {
+    const availableTools: string[] = [];
+    application.toolConfig
+      .map((tool: DefaultApplicationToolConfig) => {
+        if (tool.config.visible) {
+          availableTools.push(tool.name);
+        };
+      });
+    store.dispatch(setSelectedKey(availableTools)
```
Comment (dnlkoch) [inline src/index.tsx:189]: Why do we set the _available_ tools as _selected_ keys?

File: src/store/toolMenu/index.ts [hunk 4]
```diff
@@ -8,15 +8,16 @@ export interface ToolMenuState {
 }
 
 const initialState: ToolMenuState = {
-  selectedKeys: []
+  // This is the default state. If no config is given, then load all tools
+  selectedKeys: ['default']
```
Comment (dnlkoch) [inline src/store/toolMenu/index.ts:12]: As far as I understand this state member contains (or should contain) the currently _selected_ keys in the tool menu, not the available ones.

File: src/components/ToolMenu/index.tsx [hunk 5]
```diff
@@ -206,12 +206,15 @@ export const ToolMenu: React.FC<ToolMenuProps> = ({
 
   const onSubmenuTitleClick = (key: TitleEventEntity) => {
     setCollapsed(false);
-    selectedKeys.forEach(selKey => dispatch(unsetSelectedKey(selKey)));
+    // WHY ???
+    // toolMenu.forEach(selKey => dispatch(unsetSelectedKey(selKey)));
```
Comment (dnlkoch) [inline src/components/ToolMenu/index.tsx:210]: If not needed, please remove.

File: src/components/ToolMenu/index.tsx [hunk 6]
```diff
@@ -104,7 +104,7 @@ export const ToolMenu: React.FC<ToolMenuProps> = ({
   const map = useMap();
 
   const dispatch = useAppDispatch();
-  const selectedKeys = useAppSelector(state => state.toolMenu.selectedKeys);
+  const toolMenu = useAppSelector(state => state.toolMenu.selectedKeys);
```
Comment (dnlkoch) [inline src/components/ToolMenu/index.tsx:107]: For me this naming is misleading as the state should contain the selected keys.

File: src/store/toolMenu/index.ts (hunk 4 above)
Comment (Unraveler) [reply src/store/toolMenu/index.ts:12]: I guess I missunderstood this. I thought this were here for the available tools. I Will redo this part and create a new store entry for the tools that should render

File: src/components/ToolMenu/index.tsx (hunk 5 above)
Comment (Unraveler) [reply src/components/ToolMenu/index.tsx:210]: see comment above

File: src/index.tsx [hunk 7]
```diff
@@ -173,6 +177,22 @@ const setApplicationToStore = async (application?: Application) => {
     // @ts-ignore
     store.dispatch(setLogoPath(application.clientConfig.theme.logoPath));
   }
+
+  if (application.toolConfig) {
+    const availableTools: string[] = [];
+    application.toolConfig
+      // @ts-ignore
```
Comment (dnlkoch) [inline src/index.tsx:184]: Why do we need the `@ts-ignore`?

File: src/index.tsx [hunk 8]
```diff
@@ -173,6 +177,22 @@ const setApplicationToStore = async (application?: Application) => {
...
+      .map((tool: DefaultApplicationToolConfig) => {
+        if (tool.config.visible) {
+          availableTools.push(tool.name);
+        };
+      });
+
+    store.dispatch(setAvailableTools(availableTools)
+    );
```
Comment (dnlkoch) [inline src/index.tsx:192]:
```suggestion
    store.dispatch(setAvailableTools(availableTools));
```

File: src/index.tsx (hunk 7 above)
Comment (Unraveler) [reply src/index.tsx:184]: yes, at least while this change is not merged and available.

https://github.com/terrestris/shogun-util/pull/88

**Review by dnlkoch — APPROVED** [review]

**Pull request merged**